## Question 3: Generate Predictions (5 Marks)
Load the saved model and predict on the test set (hold-out from train-test split). Save as Mugimba predictions.csv.

In [1]:
import joblib
import pandas as pd
from sklearn.model_selection import train_test_split  # If needed to recreate split

# Load cleaned data (from Part A)
df = pd.read_csv(f'../data/Mugimba.csv')

# Assuming selected features from Question 1; adjust if yours differ
X = df.drop(['Outcome','Insulin'], axis=1)
y = df['Outcome']

# Recreate test set if not already (use same random_state as in Question 2)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Load the saved model (update path/filename if renamed to 'Mugimba best model.pkl')
model = joblib.load('../models/Mugimba_best_model.pkl')

# Generate predictions
predictions = model.predict(X_test)

# Save as CSV (include actual for reference, but exam requires just predictions)
pd.DataFrame({'Predictions': predictions}).to_csv(f'../data/Mugimba_predictions.csv', index=False)
print(f"Predictions saved as '../data/Mugimba_predictions.csv'")


Predictions saved as '../data/Mugimba_predictions.csv'


In [2]:
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix
print("Accuracy:", accuracy_score(y_test, predictions))
print("F1 Score:", f1_score(y_test, predictions))
print("Confusion Matrix:\n", confusion_matrix(y_test, predictions))

Accuracy: 0.6716417910447762
F1 Score: 0.5
Confusion Matrix:
 [[34 13]
 [ 9 11]]


Question 4: Interpretation of Model Performance (10 Marks)
Evaluation Metrics
The best model (Logistic Regression tuned) achieved the following performance on the 20% hold-out test set:

Accuracy: 67.16%
F1 Score: 50.00%
Precision: 45.83%
Recall: 55.00%

Confusion Matrix
[[34  13]
[ 9  11]]
Sources of Prediction Error

Class Imbalance (from EDA): ~65% No-Diabetes vs 35% Diabetes → model tends to predict the majority class, resulting in higher false negatives (missed diabetes cases).
Data Quality Issues: Even after imputation and outlier removal, the original dataset contained many zeros (especially Insulin and SkinThickness), which introduced noise.
Limited Features: Only 8 clinical variables; missing important factors such as family history text notes, lifestyle data, or HbA1c levels.
Model Limitations: Logistic Regression assumes linear relationships; non-linear interactions (e.g., Glucose × BMI) are not fully captured.

Practical Implications in Healthcare (Patient Intelligence)

False Negatives (missed high-risk patients) are the most dangerous → a diabetic patient may not receive timely intervention, leading to complications (e.g., retinopathy, kidney failure) and higher long-term costs.
False Positives → unnecessary tests or medication, increasing hospital workload and patient anxiety.
Real-World Deployment: The model should be used as a risk-scoring tool (not final diagnosis) with a lower decision threshold (e.g., probability > 0.4) to increase recall and catch more at-risk patients.
Recommendation: Combine with textual clinical notes (Part C) and retrain periodically to reduce model drift.

Overall, while the model provides valuable predictive intelligence, it must be integrated carefully into clinical workflows with human oversight to minimise patient harm.